msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/OnBehalfOfTests.java

```diff
@@ -0,0 +1,95 @@
+import java.util.Map;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.junit.jupiter.MockitoExtension;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+import static org.mockito.Mockito.times;
+
+@ExtendWith(MockitoExtension.class)
+class OnBehalfOfTests {
+
+    private String getSuccessfulResponse(String accessToken) {
+        return "{\"access_token\":\""+accessToken+"\",\"expires_in\": \""+ 60*60*1000 +"\",\"token_type\":" +
+                "\"Bearer\",\"client_id\":\"client_id\",\"Content-Type\":\"text/html; charset=utf-8\"}";
+    }
+
+    private HttpResponse expectedResponse(int statusCode, String response) {
+        Map<String, List<String>> headers = new HashMap<String, List<String>>();
+        headers.put("Content-Type", Collections.singletonList("application/json"));
+
+        HttpResponse httpResponse = new HttpResponse();
+        httpResponse.statusCode(statusCode);
+        httpResponse.body(response);
+        httpResponse.addHeaders(headers);
+
+        return httpResponse;
+    }
+
+    @Test
+    void OnBehalfOf_InternalCacheLookup_Success() throws Exception {
+        DefaultHttpClient httpClientMock = mock(DefaultHttpClient.class);
+
+        when(httpClientMock.send(any(HttpRequest.class))).thenReturn(expectedResponse(200, getSuccessfulResponse("token")));
+
+        ConfidentialClientApplication cca =
+                ConfidentialClientApplication.builder("clientId", ClientCredentialFactory.createFromSecret("password"))
+                                .authority("https://login.microsoftonline.com/tenant/")
+                        .instanceDiscovery(false)
+                        .validateAuthority(false)
+                        .httpClient(httpClientMock)
+                        .build();
+
+        OnBehalfOfParameters parameters = OnBehalfOfParameters.builder(Collections.singleton("scopes"), new UserAssertion(TestHelper.signedToken)).build();
+
+        IAuthenticationResult result = cca.acquireToken(parameters).get();
+        IAuthenticationResult result2 = cca.acquireToken(parameters).get();
+
+        //OBO flow should perform an internal cache lookup, so similar parameters should only cause one HTTP client call
+        assertEquals(result.accessToken(), result2.accessToken());
+        verify(httpClientMock, times(1)).send(any());
+    }
+
+    @Test
+    void OnBehalfOf_TenantOverride() throws Exception {
+        DefaultHttpClient httpClientMock = mock(DefaultHttpClient.class);
+
+        ConfidentialClientApplication cca =
+                ConfidentialClientApplication.builder("clientId", ClientCredentialFactory.createFromSecret("password"))
+                        .authority("https://login.microsoftonline.com/tenant")
+                        .instanceDiscovery(false)
```

Comment:
ok for now, but the http mock can easily handle this as well.

Reply:
This was mainly a quick fix for counting the number of times I made a call to the token endpoint vs. the cache: I wasn't testing instance discovery anyway, and if I did do instance discovery I'd need to both mock it and count an extra call in every `verify(httpClientMock, times(...)).send()` check

It's definitely better to mock it and will set that up in a future PR, but I wanted to keep this PR focused on the new OBO tests and essentially be a proof-of-concept for how to improve the unit tests.